walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,185 +11,354 @@
-    int windowClose = 0; 
-
-    Object[] objectArray = events.toArray(); 
-    Event[] eventsOrderedByStart = new Event[objectArray.length]; 
-    Event[] eventsOrderedByEnd = new Event[objectArray.length];
-
-    for (int i = 0; i < events.toArray().length; i++) {
-      eventsOrderedByStart[i] = (Event) objectArray[i];  
-      eventsOrderedByEnd[i] = (Event) objectArray[i]; 
+  // Method for removing the overlapping times from our returned suggested times 
+  private void removeOverlaps() {
+    ArrayList<TimeRange> duplicatesToRemove = new ArrayList<>();
+    for (TimeRange time: optimalTimesForMeeting) {
+      for (TimeRange otherTime: optimalTimesForMeeting) {
+        if (time == otherTime) continue;
+        if (time.contains(otherTime)) {
+          duplicatesToRemove.add(otherTime);
+        }
+      }
     }
+    optimalTimesForMeeting.removeAll(duplicatesToRemove);
+  }
 
-    Arrays.sort(eventsOrderedByStart, ORDER_EVENT_BY_START);
-    Arrays.sort(eventsOrderedByEnd, ORDER_EVENT_BY_END);
-    int startPointer = 0; 
-    int endPointer = 0; 
-    int nextMeetingEnd = 0; 
-    int nextMeetingStart = 0; 
-    Set<Event> problemEvents = new HashSet<>();
-
-
-
-    // Loop until all meetings have ended
-    while(endPointer < eventsOrderedByEnd.length) {
-    // What comes first, the end of a meeting or the start of one
-      nextMeetingEnd = eventsOrderedByEnd[endPointer].getWhen().end(); 
-      if (startPointer < eventsOrderedByStart.length) {
-      nextMeetingStart = eventsOrderedByStart[startPointer].getWhen().start(); 
-      }
-      if (startPointer == eventsOrderedByStart.length || nextMeetingEnd <= nextMeetingStart) {
+  private void findPotentialTimesForRequiredAttendees() {
+    int windowStart = 0;
+    int windowClose = 0;
+    startPointer = 0;
+    endPointer = 0;
+    nextMeetingEndTime = 0;
+    nextMeetingStartTime = 0;
+    // problemEvents keeps track of the events which have relevant guests
+    Set<Event> problemEvents = new HashSet<>(); 
+    while (endPointer < eventsOrderedByEnd.length) {
+      if (oldMeetingEndBeforeNewStart()) {
         if (problemEvents.contains(eventsOrderedByEnd[endPointer])) {
-          windowStart = nextMeetingEnd; 
-          problemEvents.remove(eventsOrderedByEnd[endPointer]); 
+          windowStart = nextMeetingEndTime;
+          problemEvents.remove(eventsOrderedByEnd[endPointer]);
         }
-        endPointer++; 
-      } else {
-          if (isEventImportantWithOptional(eventsOrderedByStart[startPointer], request)) {
-            windowClose = nextMeetingStart; 
-            if (problemEvents.isEmpty()) {
-              if ((windowClose - windowStart) >= duration) {
-                result.add(TimeRange.fromStartEnd(windowStart, windowClose, false)); 
-              }
-            }
-            problemEvents.add(eventsOrderedByStart[startPointer]); 
+        endPointer++;
+        continue;
+      }
+      if (isEventImportant(eventsOrderedByStart[startPointer])) {
```

Comment:
```
if (isEventImportant(eventsOrderedByStart[startPointer]) && problemEvents.isEmpty() &&  (nextMeetingStartTime - windowStart) >= request.getDuration())
```

 instead of the nested if.

Reply:
Okay, just made the change.